internal/proxy/task_search.go

```diff
@@ -455,38 +480,241 @@ func (t *searchTask) initAdvancedSearchRequest(ctx context.Context) error {
 	if err != nil {
+		log.Warn("failed to get primary field schema", zap.Error(err))
+		return err
+	}
+	if t.result, err = rankSearchResultData(ctx, t.SearchRequest.GetNq(),
+		t.rankParams,
+		primaryFieldSchema.GetDataType(),
+		multipleMilvusResults,
+		t.SearchRequest.GetGroupByFieldId(),
+		t.SearchRequest.GetGroupSize(),
+		t.groupScorer); err != nil {
+		log.Warn("rank search result failed", zap.Error(err))
 		return err
 	}
-	t.groupScorer = groupScorer
+
+	if t.needRequery {
+		if t.requery == nil {
+			t.requery = requery
+		}
+		queryResult, err := t.requery(t, span, t.result.Results.Ids, t.translatedOutputFields)
+		if err != nil {
+			log.Warn("failed to requery", zap.Error(err))
+			return err
+		}
+		fields, err := t.reorganizeRequeryResults(ctx, queryResult, []*schemapb.IDs{t.result.Results.Ids})
+		if err != nil {
+			return err
+		}
+		t.result.Results.FieldsData = fields[0]
+	}
 
 	return nil
 }
 
+func mergeIDs(idsList []*schemapb.IDs) (*schemapb.IDs, int) {
+	uniqueIDs := &schemapb.IDs{}
+	count := 0
+	switch idsList[0].GetIdField().(type) {
+	case *schemapb.IDs_IntId:
+		idsSet := typeutil.NewSet[int64]()
+		for _, ids := range idsList {
+			if data := ids.GetIntId().GetData(); data != nil {
+				idsSet.Insert(data...)
+			}
+		}
+		count = idsSet.Len()
+		uniqueIDs.IdField = &schemapb.IDs_IntId{
+			IntId: &schemapb.LongArray{
+				Data: idsSet.Collect(),
+			},
+		}
+	case *schemapb.IDs_StrId:
+		idsSet := typeutil.NewSet[string]()
+		for _, ids := range idsList {
+			if data := ids.GetStrId().GetData(); data != nil {
+				idsSet.Insert(data...)
+			}
+		}
+		count = idsSet.Len()
+		uniqueIDs.IdField = &schemapb.IDs_StrId{
+			StrId: &schemapb.StringArray{
+				Data: idsSet.Collect(),
+			},
+		}
+	}
+	return uniqueIDs, count
+}
+
+func (t *searchTask) hybridSearchRank(ctx context.Context, span trace.Span, multipleMilvusResults []*milvuspb.SearchResults) error {
+	var err error
+	// The first step of hybrid search is without meta information. If rerank requires meta data, we need to do requery.
+	// At this time, outputFields and rerank input_fields will be recalled.
+	// If we want to save memory, we can only recall the rerank input_fields in this step, and recall the output_fields in the third step
+	if t.needRequery {
+		idsList := lo.FilterMap(multipleMilvusResults, func(m *milvuspb.SearchResults, _ int) (*schemapb.IDs, bool) {
+			return m.Results.Ids, true
+		})
+		allIDs, count := mergeIDs(idsList)
+		if count == 0 {
+			t.result = &milvuspb.SearchResults{
+				Status: merr.Success(),
+				Results: &schemapb.SearchResultData{
+					NumQueries: t.Nq,
+					TopK:       t.rankParams.limit,
+					FieldsData: make([]*schemapb.FieldData, 0),
+					Scores:     []float32{},
+					Ids:        &schemapb.IDs{},
+					Topks:      []int64{},
+				},
+			}
+			return nil
+		}
+		allNames := typeutil.NewSet[string](t.translatedOutputFields...)
+		allNames.Insert(t.functionScore.GetAllInputFieldNames()...)
```

Comment:
Plz correct me if I am wrong, we only do 2-phases (search ,requery before rerank) instead of 3(search, requery rerank-needed column, requery output)

Reply:
Two more Qs:
1. Can we ensure function input can cover all the output user might sepcify?
2. Why we use field name instead of id?